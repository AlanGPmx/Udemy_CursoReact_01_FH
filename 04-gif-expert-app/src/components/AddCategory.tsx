import {useState} from 'react';
import {AddCategoryInterface} from '../interfaces/AddCategory.interface';
import PropTypes from 'prop-types';

export const AddCategory = (props: AddCategoryInterface) => {
	const {onNewCategory} = props;
	const [inputValue, setInputValue] = useState('');

	const onInputChange = (event: any) => {
		const {target} = event;
		const {value} = target;
		setInputValue(value);
	};

	const onSubmit = (event: any) => {
		event.preventDefault();
		const newValue = inputValue.trim();

		if (newValue.length <= 1) return;

		setInputValue('');
		onNewCategory(newValue);
	};

	return (
		<form onSubmit={onSubmit} aria-label='form'>
			<input
				type='text'
				placeholder='Busca un GIF...'
				value={inputValue}
				onChange={onInputChange}
			/>
		</form>
	);
};

AddCategory.propTypes = {
	onNewCategory: PropTypes.func.isRequired,
};
