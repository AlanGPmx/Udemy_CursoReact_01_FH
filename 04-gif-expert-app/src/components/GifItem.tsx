import {GifItemInterface} from '../interfaces/GiftItem.interface';
import PropTypes from 'prop-types';

export const GifItem = (props: GifItemInterface) => {
	const {title, url} = props;
	return (
		<div className='card'>
			<img src={url} alt={title} />
			<p>{title}</p>
		</div>
	);
};

// Es la alternativa en JS y React al tipado que TS ofrece
GifItem.propTypes = {
	title: PropTypes.string.isRequired,
	url: PropTypes.string.isRequired,
};

//La alternativa para asignar un valor por defecto a las Props y no hacerlo en crudo en los argumentos del componente
GifItem.defaultProps = {
	url: 'https://www.publicdomainpictures.net/pictures/280000/nahled/not-found-image-15383864787lu.jpg',
	title: 'Imagen no encontrada',
};
