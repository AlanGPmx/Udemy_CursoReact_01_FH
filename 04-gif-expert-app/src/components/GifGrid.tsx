import {useFetchGifs} from '../hooks/useFetchGifs';
import {GifGridInterface} from '../interfaces/GifGridInterface';
import {GifItem} from './GifItem';
import PropTypes from 'prop-types';

export const GifGrid = (props: GifGridInterface) => {
	const {category} = props;

	const {gifs, isLoading} = useFetchGifs(category);

	return (
		<>
			<h3>{category}</h3>
			{isLoading && <h2>Cargando...</h2>}
			<div className='card-grid'>
				{gifs.map((_gif: any) => (
					<GifItem key={_gif.id} {..._gif} />
				))}
			</div>
		</>
	);
};

GifGrid.propTypes = {
	category: PropTypes.string.isRequired,
};
