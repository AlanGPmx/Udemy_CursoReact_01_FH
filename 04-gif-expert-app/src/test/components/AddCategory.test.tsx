import {fireEvent, render, screen} from '@testing-library/react';
import {AddCategory} from '../../components';

describe('Pruebas en <AddCategory />', () => {
	const onNewCategory = jest.fn();

	test('Deberá renderizar el componente y compararlo con el Snapshot', () => {
		const {container} = render(<AddCategory onNewCategory={onNewCategory} />);
		expect(container).toMatchSnapshot();
	});

	test('Deberá cambiar el valor de la caja de texto con nueva categoria', () => {
		render(<AddCategory onNewCategory={onNewCategory} />);
		const newCategoryToAdd = 'Hola Mundo';
		const input = screen.getByRole('textbox') as HTMLInputElement;
		fireEvent.change(input, {target: {value: newCategoryToAdd}});

		expect(input.value).toBe(newCategoryToAdd);
	});

	test('Debe llamar onNewCategory si el input tiene un valor', () => {
		render(<AddCategory onNewCategory={onNewCategory} />);
		const inputValue = 'Hola Mundo';
		const input = screen.getByRole('textbox') as HTMLInputElement;
		const form = screen.getByRole('form');

		fireEvent.change(input, {target: {value: inputValue}});
		fireEvent.submit(form);

		expect(input.value).toBe('');
		expect(onNewCategory).toHaveBeenCalled(); // Puede ser: expect(onNewCategory).toHaveBeenCalledTimes(1);
		expect(onNewCategory).toHaveBeenCalledWith(inputValue);
	});

	test('No debe llamar onNewCategory si el input tiene esta vacio', () => {
		render(<AddCategory onNewCategory={onNewCategory} />);
		const form = screen.getByRole('form');

		fireEvent.submit(form);

		expect(onNewCategory).toHaveBeenCalledTimes(0);
		expect(onNewCategory).not.toHaveBeenCalled(); // Segunda alternativa
	});
});
