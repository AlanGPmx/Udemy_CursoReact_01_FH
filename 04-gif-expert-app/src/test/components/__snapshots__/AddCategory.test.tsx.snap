// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Deber√° renderizar el componente y compararlo con el Snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Busca un GIF..."
      type="text"
      value=""
    />
  </form>
</div>
`;
